torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -353,119 +353,137 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+    return sz;
+  });
+  pickler.protocol();
+  pickler.pushIValue(buffers.tensors);
+  pickler.stop();
+  tpMessage.payloads.push_back(tensorpipe::Message::Payload{
+      buffers.pickle.data(), buffers.pickle.size()});
+  for (const auto& tensorData : pickler.tensorData()) {
     // Enforce memory copy if tensor is created from torch::from_blob, means
     // that the tensor doesn't own the memory.
     if (!tensorData.storageHasDeleter()) {
-      std::vector<uint8_t> storageData(tensorData.sizeInBytes());
-      memcpy(storageData.data(), tensorData.data(), storageData.size());
-      tpTensor.data = storageData.data();
-      copiedTensors.push_back(std::move(storageData));
+      std::vector<char> storageData(
+          tensorData.data(), tensorData.data() + tensorData.sizeInBytes());
+      tpMessage.tensors.push_back(
+          tensorpipe::Message::Tensor{storageData.data(), storageData.size()});
+      buffers.copiedTensors.push_back(std::move(storageData));
     } else {
-      tpTensor.data = (uint8_t*)(tensorData.data());
+      tpMessage.tensors.push_back(tensorpipe::Message::Tensor{
+          const_cast<char*>(tensorData.data()), tensorData.sizeInBytes()});
     }
-    tpTensor.length = tensorData.sizeInBytes();
-    tpMessage.tensors.push_back(std::move(tpTensor));
   }
 
-  return TensorPipeEntry{std::move(tpMessage),
-                         std::move(reservedTensors),
-                         std::move(copiedTensors)};
+  return std::make_tuple(std::move(tpMessage), std::move(buffers));
 }
 
-Message tensorpipeAllocateMessage(tensorpipe::Message& tpMessage) {
-  // Payload, message type and message id
+TensorpipeReadBuffers tensorpipeAllocate(tensorpipe::Message& tpMessage) {
+  TensorpipeReadBuffers buffers;
+
   TORCH_INTERNAL_ASSERT(
-      tpMessage.payloads.size() == 1,
-      "message expected to contain 1 payload, whereas it contained ",
+      tpMessage.payloads.size() == 4,
+      "message expected to contain 4 payloads, whereas it contained ",
       tpMessage.payloads.size(),
       " payloads");
-  std::vector<char> payload(tpMessage.payloads[0].length);
-  tpMessage.payloads[0].data = (uint8_t*)(payload.data());
+
   TORCH_INTERNAL_ASSERT(
-      tpMessage.metadata.size() == 2 * sizeof(int64_t),
-      "message metadata must be ",
-      2 * sizeof(int64_t),
-      " bytes, whereas it is ",
-      tpMessage.metadata.size(),
+      tpMessage.payloads[0].length == sizeof(MessageType),
+      "first payload expected to contain ",
+      sizeof(MessageType),
+      " bytes, whereas it contained ",
+      tpMessage.payloads[0].length,
       " bytes");
-  int64_t mtypei, mId;
-  memcpy(&mtypei, tpMessage.metadata.data(), sizeof(int64_t));
-  memcpy(&mId, tpMessage.metadata.data() + sizeof(int64_t), sizeof(int64_t));
-  MessageType mType = static_cast<MessageType>(mtypei);
+  buffers.type = std::make_unique<MessageType>();
+  tpMessage.payloads[0].data = buffers.type.get();
```

Comment:
Could you please elaborate more on this change? It looks like we no longer expect `tpMessage` to contain metadata?

Reply:
This unfortunately stems from a bad naming in TensorPipe: what is called "metadata" in the tp::Message should really have been called something like "allocation hints", i.e., it should only contain information needed in order to correctly allocate the receiver's buffers. It should not be used to store what the agent considers to be a rpc::Message's metadata, but instead it should store the metadata that corresponds to tp::Message's data. The rpc::Message type and id are just some additional payload from TensorPipe's perspective, and should be stored as such: payloads are transferred more efficiently than metadata (even though here we're just talking about a few bytes...).

The moment when we'll really need to use the "allocation hints" is for GPU tensors, as in that case the allocation hints will contain the device, which we need to know when allocating buffers